### PR TITLE
Compile primitive at

### PR DESCRIPTION
**Pull request by PalumboN**
Changes to compile `primitiveAt` on Druid:

- Avoid / ignore some format cases.

There are also:

- Clean duplicated code
- Improve MCDebugger
- Minor fixes

### CODE REVIEW

File: smalltalksrc/VMMaker/StackInterpreter.class.st [hunk 1]
```diff
@@ -14989,7 +14990,7 @@ StackInterpreter >> subscript: array with: index storing: oopToStore format: fmt
 			 objectMemory storeShort16: index - 1 ofObject: array withValue: signedValueToStore.
 			 ^self].
 		 fmt = objectMemory sixtyFourBitIndexableFormat ifTrue:
-			[| unsigned64BitValueToStore |
+			[
```
Comment (guillep) [inline smalltalksrc/VMMaker/StackInterpreter.class.st:14993]: Why this change?

Comment (guillep) [reply smalltalksrc/VMMaker/StackInterpreter.class.st:14993]: Personally, I like temporary definitions close to their usages, but maybe you have a stronger reason :)

File: smalltalksrc/VMMakerTests/UnicornARMv8Simulator.class.st [hunk 2]
```diff
@@ -221,7 +221,7 @@ UnicornARMv8Simulator >> receiverRegister [
 { #category : #accessing }
 UnicornARMv8Simulator >> registerList [
 
-	^ #(lr pc sp fp x28 x0 x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x16 x19 x20 x21 x22 x23 x24 x25 x26 zero negative carry overflow v0 v1 v2)
+	^ #(lr pc sp fp x28 x0 x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x16 x19 x20 x21 x22 x23 x24 x25 x26 x27 zero negative carry overflow v0 v1 v2)
```
Comment (guillep) [inline smalltalksrc/VMMakerTests/UnicornARMv8Simulator.class.st:224]: ;)

File: smalltalksrc/VMMakerTests/VMJittedPrimitivesTest.class.st [hunk 3]
```diff
@@ -97,25 +97,6 @@ VMJittedPrimitivesTest >> new64BitIndexableOfSize: aSize [
...
-	| padding numberOfWordSizeSlots desiredByteSize |
-	desiredByteSize := aSize * bytesPerSlot roundUpTo: self wordSize.
-	numberOfWordSizeSlots := desiredByteSize / self wordSize.
-	padding := desiredByteSize / bytesPerSlot - aSize.
-	^ memory
-		  allocateNewSpaceSlots: numberOfWordSizeSlots
-		  format: format + padding
-		  classIndex: self nextOrdinaryClassIndex
-]
-
```
Comment (guillep) [inline smalltalksrc/VMMakerTests/VMJittedPrimitivesTest.class.st:118]: Ok, IIRC, these methods are duplicated in the superclass so removing them here is safe.

**Review by guillep — APPROVED** [review]
Thanks!!

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 4]
```diff
@@ -452,8 +452,7 @@ DruidJIT class >> bytecodeTable [
 		  #isMapped. #v4:Long:Branch:Distance: }.
 		  { 2. 239. 239. #gen_ExtJumpIfFalse. #branch. #isBranchFalse.
 		  #isMapped. #v4:Long:Branch:Distance: }.
-		  { 2. 240. 240. #gen_ExtStoreAndPopReceiverVariableBytecode.
-		  #isInstVarRef. #isMappedIfImmutability }.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:456]: Why this one got unmapped?

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 5]
```diff
@@ -3878,15 +3878,15 @@ DruidJIT >> gen_PrimitiveAdd [
 	currentBlock := self Label.
 	jump1 jmpTarget: currentBlock.
 	jump2 jmpTarget: currentBlock.
-	^ CompletePrimitive
+	^ 0
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:3881]: Yes

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 6]
```diff
@@ -3878,15 +3878,15 @@ DruidJIT >> gen_PrimitiveAdd [
...
 { #category : #generated }
 DruidJIT >> gen_PrimitiveAt [
 	"AutoGenerated by Druid"
 
-	| jump1 jump7 jumpNext jump9 s68 jump3 jump6 s119 currentBlock s12 jump8 jump5 jumpTrue s162 s101 jump2 s150 jump4 |
-	self SubCq: 16 R: SPReg.
+	| jump1 s100 jumpNext s68 jump3 jump6 s160 s148 currentBlock s12 jump8 jump5 jumpTrue jump2 s118 jump7 jump4 |
+	self SubCq: 24 R: SPReg.
 	self MoveR: ReceiverResultReg R: Extra3Reg.
 	self MoveR: Extra3Reg Mw: 0 r: SPReg.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:3891]: @PalumboN Task for Juan Ignacio?

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 7]
```diff
@@ -3895,257 +3895,263 @@ DruidJIT >> gen_PrimitiveAt [
...
 	jumpTrue := self JumpZero: 0.
-	self MoveCq: 0 R: Extra0Reg.
+	self MoveCq: 0 R: Extra1Reg.
 	jumpNext := self Jump: 0.
 	jumpTrue jmpTarget: self Label.
-	self MoveCq: 1 R: Extra0Reg.
+	self MoveCq: 1 R: Extra1Reg.
 	jumpNext jmpTarget: self Label.
-	self CmpCq: 0 R: SendNumArgsReg.
+	self CmpCq: 0 R: Extra0Reg.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:3908]: @PalumboN Task for Juan Ignacio

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 8]
```diff
@@ -3895,257 +3895,263 @@ DruidJIT >> gen_PrimitiveAt [
...
+	self CmpCq: 9 R: Extra0Reg.
 	jump4 := self JumpNonZero: 0.
-	self MoveR: Extra1Reg R: TempReg.
+	self MoveR: Extra2Reg R: SendNumArgsReg.
 	jump8 := self Jump: 0.
 	currentBlock := self Label.
 	jump4 jmpTarget: currentBlock.
 	s68 := 0.
-	self MoveCq: s68 R: TempReg.
+	self MoveCq: s68 R: SendNumArgsReg.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:4001]: @PalumboN For Juan Ignacio, maybe a more advanced task? This is probably a constant in a Phi that cannot be propagated. But maybe here there is good reason for tail duplication?

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 9]
```diff
@@ -3895,257 +3895,263 @@ DruidJIT >> gen_PrimitiveAt [
...
-	self MoveCq: 0 R: Extra1Reg.
+	self MoveCq: 0 R: Extra2Reg.
 	jumpNext := self Jump: 0.
 	jumpTrue jmpTarget: self Label.
-	self MoveCq: 1 R: Extra1Reg.
+	self MoveCq: 1 R: Extra2Reg.
 	jumpNext jmpTarget: self Label.
-	self CmpCq: 9 R: SendNumArgsReg.
+	self CmpCq: 9 R: Extra0Reg.
 	jump8 := self JumpBelow: 0.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:4014]: @PalumboN this is fishy here, it's doing the same comparison twice. Maybe a nice semi advanced issue for Juan Ignacio.

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 10]
```diff
@@ -4384,7 +4383,7 @@ DruidJIT >> gen_PrimitiveEqual [
 	self genPrimReturn.
 	currentBlock := self Label.
 	jump1 jmpTarget: currentBlock.
-	^ CompletePrimitive
+	^ 0
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:4386]: Also I'd like to know the reason of this 0 here :) What is the interpreter code ignored during primitiveEqual.

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 11]
```diff
@@ -5046,35 +5042,35 @@ DruidJIT >> gen_PrimitiveNewWithArg [
...
+		Quo: Extra0Reg
+		Rem: ClassReg.
 	s105 := 8.
-	self MoveCq: s105 R: Extra0Reg.
-	self SubR: SendNumArgsReg R: Extra0Reg.
-	self AndCq: 7 R: Extra0Reg.
-	self AddR: Extra0Reg R: Extra1Reg.
-	self MoveR: ClassReg R: SendNumArgsReg.
+	self MoveCq: s105 R: ClassReg.
+	self SubR: SendNumArgsReg R: ClassReg.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:5070]: @PalumboN for Juan Ignacio. Could this be collapsed in a single instruction?

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 12]
```diff
@@ -5124,20 +5120,20 @@ DruidJIT >> gen_PrimitiveNewWithArg [
...
 	jump11 jmpTarget: currentBlock.
 	self MoveMw: 0 r: SPReg R: Extra3Reg.
-	self MoveM32: 4 r: Extra3Reg R: Extra2Reg.
-	self AndCq: 16rFFFFFFFF R: Extra2Reg.
-	self AndCq: 16r3FFFFF R: Extra2Reg.
-	self CmpCq: 0 R: Extra2Reg.
+	self MoveM32: 4 r: Extra3Reg R: Extra1Reg.
+	self AndCq: 16rFFFFFFFF R: Extra1Reg.
+	self AndCq: 16r3FFFFF R: Extra1Reg.
+	self CmpCq: 0 R: Extra1Reg.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:5130]: Here we have two ANDs but the first one includes the second one. So it can be removed.
Also we could collapse logical and bit AND/OR instructions with comparisons!

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 13]
```diff
@@ -5153,79 +5149,80 @@ DruidJIT >> gen_PrimitiveNewWithArg [
...
+	self MoveCq: 0 R: SendNumArgsReg.
 	jumpNext := self Jump: 0.
 	jumpTrue jmpTarget: self Label.
-	self MoveCq: 1 R: ClassReg.
+	self MoveCq: 1 R: SendNumArgsReg.
 	jumpNext jmpTarget: self Label.
 	self MoveMw: 8 r: SPReg R: Extra3Reg.
 	self CmpCq: 16rFFFF R: Extra3Reg.
 	jump7 := self JumpBelowOrEqual: 0.
 	jump10 := self Jump: 0.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:5170]: We could also have some code layout algorithms and some heuristics of common paths

File: smalltalksrc/VMMaker/DruidJIT.class.st [hunk 14]
```diff
@@ -5153,79 +5149,80 @@ DruidJIT >> gen_PrimitiveNewWithArg [
...
 	s194 := 8.
-	self MoveCq: s194 R: ClassReg.
+	self MoveCq: s194 R: SendNumArgsReg.
 	self MoveMw: 8 r: SPReg R: Extra3Reg.
-	self MoveR: Extra3Reg R: Extra1Reg.
-	self MulR: ClassReg R: Extra1Reg.
-	self MoveR: Extra1Reg R: ClassReg.
+	self MoveR: Extra3Reg R: ClassReg.
+	self MulR: SendNumArgsReg R: ClassReg.
+	self MoveR: ClassReg R: SendNumArgsReg.
```
Comment (guillep) [inline smalltalksrc/VMMaker/DruidJIT.class.st:5209]: Here we have a multiply by 8 :) => << 3

**Review by guillep — COMMENTED** [review]
Good now we hace material :)

File: smalltalksrc/VMMaker/StackInterpreter.class.st (hunk 1 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/StackInterpreter.class.st:14993]: Haha because Druid is not-too-much smart and makes the lookup of the variable directly in the outer context. Was easier to change the variable than fix it.

I also prefer the variable in the closest scope, but the rest of the code doesn't follow it. The IDE only adds temps to the method... So blah...

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 9 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:4014]: Weird

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 6 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:3891]: https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 7 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:3908]: https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 8 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:4001]: https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 9 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:4014]: https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 10 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:4386]: https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 11 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:5070]: https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 13 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:5170]: https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 14 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:5209]: https://github.com/Alamvic/druid/issues/107

File: smalltalksrc/VMMaker/DruidJIT.class.st (hunk 12 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/DruidJIT.class.st:5130]: https://github.com/Alamvic/druid/issues/105 and https://github.com/Alamvic/druid/issues/106

File: smalltalksrc/VMMaker/StackInterpreter.class.st (hunk 1 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/StackInterpreter.class.st:14993]: https://github.com/Alamvic/druid/issues/108
